ethcore/evm/src/interpreter/mod.rs

```diff
@@ -102,127 +103,234 @@ enum InstructionResult<Gas> {
+	/// The VM has just finished execution in the current step.
+	Done(vm::Result<GasLeft>),
+	/// The VM can continue to run.
+	Continue,
+}
+
 /// Intepreter EVM implementation
 pub struct Interpreter<Cost: CostType> {
 	mem: Vec<u8>,
 	cache: Arc<SharedCache>,
+	params: InterpreterParams,
+	reader: CodeReader,
 	return_data: ReturnData,
+	informant: informant::EvmInformant,
+	do_trace: bool,
+	done: bool,
+	valid_jump_destinations: Option<Arc<BitSet>>,
+	gasometer: Gasometer<Cost>,
+	stack: VecStack<U256>,
 	_type: PhantomData<Cost>,
 }
 
 impl<Cost: CostType> vm::Vm for Interpreter<Cost> {
-	fn exec(&mut self, params: ActionParams, ext: &mut vm::Ext) -> vm::Result<GasLeft> {
-		self.mem.clear();
+	fn exec(&mut self, ext: &mut vm::Ext) -> vm::Result<GasLeft> {
+		loop {
+			let result = self.step(ext);
+			match result {
+				InterpreterResult::Continue => {},
+				InterpreterResult::Done(value) => return value,
+				InterpreterResult::Stopped => panic!("Attempted to execute an already stopped VM.")
+			}
+		}
+	}
+}
 
-		let mut informant = informant::EvmInformant::new(ext.depth());
-		let mut do_trace = true;
+impl<Cost: CostType> Interpreter<Cost> {
+	/// Create a new `Interpreter` instance with shared cache.
+	pub fn new(mut params: ActionParams, cache: Arc<SharedCache>, ext: &vm::Ext) -> vm::Result<Interpreter<Cost>> {
+		let reader = CodeReader::new(params.code.take().expect("VM always called with code; qed"));
+		let params = InterpreterParams::from(params);
+		let informant = informant::EvmInformant::new(ext.depth());
+		let valid_jump_destinations = None;
+		let gasometer = Gasometer::<Cost>::new(Cost::from_u256(params.gas)?);
+		let stack = VecStack::with_capacity(ext.schedule().stack_limit, U256::zero());
+
+		Ok(Interpreter {
+			cache, params, reader, informant,
+			valid_jump_destinations, gasometer, stack,
+			done: false,
+			do_trace: true,
+			mem: Vec::new(),
+			return_data: ReturnData::empty(),
+			_type: PhantomData,
+		})
+	}
 
-		let code = &params.code.as_ref().expect("exec always called with code; qed");
-		let mut valid_jump_destinations = None;
+	/// Execute a single step on the VM.
+	#[inline(always)]
+	pub fn step(&mut self, ext: &mut vm::Ext) -> InterpreterResult {
+		macro_rules! try_or_done {
+			( $expr: expr ) => {
+				match $expr {
+					Ok(value) => value,
+					Err(err) => return InterpreterResult::Done(Err(err)),
+				}
+			}
+		}
 
-		let mut gasometer = Gasometer::<Cost>::new(Cost::from_u256(params.gas)?);
-		let mut stack = VecStack::with_capacity(ext.schedule().stack_limit, U256::zero());
-		let mut reader = CodeReader::new(code);
+		if self.done {
+			return InterpreterResult::Stopped;
+		}
 
-		while reader.position < code.len() {
-			let opcode = code[reader.position];
-			let instruction = Instruction::from_u8(opcode);
-			reader.position += 1;
+		let result = {
+			let mut inner = || {
+				// This case is needed because code length can be zero.
+				if self.reader.position >= self.reader.len() {
```

Comment:
Seems that we are checking that twice now, can't this be moved as some initial step? The `reader.position` is changed only during regular execution and jumps, right?

Reply:
I moved it up to check just `reader.len() == 0`. That check still needs to happen every time `step` is called, otherwise we'll need an additional "interpreter state" (which might actually cost more memory).